From 2d9a00f4a641a9022e93c6b7b5e79b830e4e3d7a Mon Sep 17 00:00:00 2001
From: Alin Jerpelea <jerpelea@gmail.com>
Date: Wed, 3 Apr 2013 22:54:50 +0300
Subject: [PATCH] mogami: enable als sensor

Change-Id: Ifdfdfd89059702c5af56a273090b3f0e488b2960
---
 BoardConfigCommon.mk |    4 ++++
 1 file changed, 4 insertions(+)

diff --git a/BoardConfigCommon.mk b/BoardConfigCommon.mk
index b4bdd19..27e138e 100644
--- a/BoardConfigCommon.mk
+++ b/BoardConfigCommon.mk
@@ -85,5 +85,9 @@ BOARD_RECOVERY_BASE := 0x00200000
 TARGET_BOOTANIMATION_PRELOAD := true
 TARGET_BOOTANIMATION_TEXTURE_CACHE := true
 
+# Light sensor
+SOMC_CFG_SENSORS_LIGHT_AS3676 := yes
+SOMC_CFG_SENSORS_LIGHT_AS3676_PATH := "/sys/devices/i2c-0/0-0040/adc_als_value"
+
 # A custom ota package maker for a device without an exposed boot partition
 TARGET_RELEASETOOL_OTA_FROM_TARGET_SCRIPT := device/semc/msm7x30-common/releasetools/semc_ota_from_target_files
-- 
1.7.10.4

