diff --git a/arch/arm/mach-msm/leds-coconut.c b/arch/arm/mach-msm/leds-coconut.c
index 74890bb..92cd49a 100755
--- a/arch/arm/mach-msm/leds-coconut.c
+++ b/arch/arm/mach-msm/leds-coconut.c
@@ -23,6 +23,7 @@ static struct as3676_platform_led as3676_leds_mapping[] = {
 	{
 		.name = "button-backlight",
 		.sinks = BIT(AS3676_SINK_RGB2) | BIT(AS3676_SINK_RGB3),
+		.flags = AS3676_FLAG_ALS_GROUP2 ,
 		.max_current = 25000,
 	},
 	{
diff --git a/arch/arm/mach-msm/leds-iyokan.c b/arch/arm/mach-msm/leds-iyokan.c
index 259101a..13df1df 100644
--- a/arch/arm/mach-msm/leds-iyokan.c
+++ b/arch/arm/mach-msm/leds-iyokan.c
@@ -23,11 +23,13 @@ static struct as3676_platform_led as3676_leds_mapping[] = {
 	{
 		.name = "button-backlight",
 		.sinks = BIT(AS3676_SINK_RGB1),
+		.flags = AS3676_FLAG_ALS_GROUP2,
 		.max_current = 25000,
 	},
 	{
 		.name = "keyboard-backlight",
 		.sinks = BIT(AS3676_SINK_32) | BIT(AS3676_SINK_33),
+		.flags = AS3676_FLAG_ALS_GROUP2,
 		.max_current = 38000,
 	},
 	{
diff --git a/arch/arm/mach-msm/leds-mango.c b/arch/arm/mach-msm/leds-mango.c
index c07ee54..c85e6cf 100644
--- a/arch/arm/mach-msm/leds-mango.c
+++ b/arch/arm/mach-msm/leds-mango.c
@@ -30,6 +30,7 @@ static struct as3676_platform_led as3676_leds_mapping[] = {
 		.name = "keyboard-backlight",
 		.sinks = BIT(AS3676_SINK_30) | BIT(AS3676_SINK_31) |
 				BIT(AS3676_SINK_32) | BIT(AS3676_SINK_33),
+		.flags = AS3676_FLAG_ALS_GROUP2,
 		.max_current = 25000,
 	},
 	{
