diff --git a/prebuilt/hw_config.sh b/prebuilt/hw_config.sh
index 3e5cedd..104c605 100644
--- a/prebuilt/hw_config.sh
+++ b/prebuilt/hw_config.sh
@@ -49,7 +49,10 @@ echo $val_filter > $dev/filter  # RFilter. Valid range is 0 - 3.
 
 # LMU AS3676 Configuration
 dev=/sys/devices/i2c-0/0-0040/leds
-echo 4000 > $dev/button-backlight/max_current
+echo 2,127,1,130,130,2,2 > $dev/button-backlight/als/curve  # ALS curve for group2
+echo 2,127,1,130,130,2,2 > $dev/keyboard-backlight/als/curve  # ALS curve for group2
+echo 1 > $dev/button-backlight/als/enable  #Sensor on/off. 1 = on, reg 90h
+echo 1 > $dev/keyboard-backlight/als/enable  #Sensor on/off. 1 = on, reg 90h
 echo 38000 > $dev/keyboard-backlight/max_current
 echo 10000 >$dev/tally-light/max_current
 
