diff --git a/prebuilt/hw_config.sh b/prebuilt/hw_config.sh
index c7d63fb..a3389bd 100644
--- a/prebuilt/hw_config.sh
+++ b/prebuilt/hw_config.sh
@@ -48,6 +48,8 @@ echo $val_filter > $dev/filter  # RFilter. Valid range is 0 - 3.
 
 # LMU AS3676 Configuration
 dev=/sys/devices/i2c-0/0-0040/leds
+echo 2,48,0,128,128,4,10 > $dev/button-backlight/als/curve  # ALS curve for group2
+echo 1 > $dev/button-backlight/als/enable  #Sensor on/off. 1 = on, reg 90h
 echo 500 > $dev/button-backlight/max_current
 
 # TI BQ275xx firmware loader
