diff --git a/core/java/android/server/BluetoothDeviceProperties.java b/core/java/android/server/BluetoothDeviceProperties.java
index fe3ef79..7d5f0b4 100644
--- a/core/java/android/server/BluetoothDeviceProperties.java
+++ b/core/java/android/server/BluetoothDeviceProperties.java
@@ -58,16 +58,37 @@ class BluetoothDeviceProperties {
                 }
                 if (name.equals("UUIDs") || name.equals("Nodes")) {
                     StringBuilder str = new StringBuilder();
-                    len = Integer.valueOf(properties[++i]);
-                    for (int j = 0; j < len; j++) {
-                        str.append(properties[++i]);
-                        str.append(",");
-                    }
-                    if (len > 0) {
-                        newValue = str.toString();
+                    i++;
+                    if (i < properties.length) {
+                        len = Integer.valueOf(properties[i]);
+                        for (int j = 0; j < len; j++) {
+                            i++;
+                            if (i < properties.length) {
+                                str.append(properties[i]);
+                                str.append(",");
+                            } else {
+                                Log.e(TAG, "Error: Remote Device Property Value at index "
+                                    + i + " is missing");
+                                continue;
+                            }
+                        }
+                        if (len > 0) {
+                            newValue = str.toString();
+                        }
+                    } else {
+                        Log.e(TAG, "Error: Remote Device Property Value at index "
+                            + i + " is missing");
+                        continue;
                     }
                 } else {
-                    newValue = properties[++i];
+                    i++;
+                    if (i < properties.length) {
+                        newValue = properties[i];
+                    } else {
+                        Log.e(TAG, "Error: Remote Device Property Value at index "
+                            + i + " is missing");
+                        continue;
+                    }
                 }
 
                 propertyValues.put(name, newValue);
