diff --git a/net/compat-wireless/drivers/net/wireless/wl12xx/main.c b/net/compat-wireless/drivers/net/wireless/wl12xx/main.c
index 27cde0c..67c1b08 100644
--- a/net/compat-wireless/drivers/net/wireless/wl12xx/main.c
+++ b/net/compat-wireless/drivers/net/wireless/wl12xx/main.c
@@ -5241,7 +5241,6 @@ int wl1271_init_ieee80211(struct wl1271 *wl)
 
 	wl->hw->flags = IEEE80211_HW_SIGNAL_DBM |
 		IEEE80211_HW_BEACON_FILTER |
-		IEEE80211_HW_SUPPORTS_PS |
 		IEEE80211_HW_SUPPORTS_UAPSD |
 		IEEE80211_HW_HAS_RATE_CONTROL |
 		IEEE80211_HW_CONNECTION_MONITOR |
