diff --git a/arch/arm/mach-msm/qdsp5v2/snddev_data_timpani.c b/arch/arm/mach-msm/qdsp5v2/snddev_data_timpani.c
index 68666b7..8ec84e9 100644
--- a/arch/arm/mach-msm/qdsp5v2/snddev_data_timpani.c
+++ b/arch/arm/mach-msm/qdsp5v2/snddev_data_timpani.c
@@ -74,10 +74,10 @@ static struct snddev_icodec_data snddev_iearpiece_ffa_data = {
 	.default_sample_rate = 48000,
 	.pamp_on = NULL,
 	.pamp_off = NULL,
-	.max_voice_rx_vol[VOC_NB_INDEX] = -700,
-	.min_voice_rx_vol[VOC_NB_INDEX] = -2200,
-	.max_voice_rx_vol[VOC_WB_INDEX] = -1400,
-	.min_voice_rx_vol[VOC_WB_INDEX] = -2900,
+	.max_voice_rx_vol[VOC_NB_INDEX] = 1000, //700,
+	.min_voice_rx_vol[VOC_NB_INDEX] = -1900, //-2200,
+	.max_voice_rx_vol[VOC_WB_INDEX] = -1100, //-1400,
+	.min_voice_rx_vol[VOC_WB_INDEX] = -2600, //-2900,
 };
 
 static struct platform_device msm_iearpiece_ffa_device = {
@@ -233,10 +233,10 @@ static struct snddev_ecodec_data snddev_bt_sco_earpiece_data = {
 	.conf_pcm_ctl_val = BT_SCO_PCM_CTL_VAL,
 	.conf_aux_codec_intf = BT_SCO_AUX_CODEC_INTF,
 	.conf_data_format_padding_val = BT_SCO_DATA_FORMAT_PADDING,
-	.max_voice_rx_vol[VOC_NB_INDEX] = 400,
-	.min_voice_rx_vol[VOC_NB_INDEX] = -1100,
-	.max_voice_rx_vol[VOC_WB_INDEX] = 400,
-	.min_voice_rx_vol[VOC_WB_INDEX] = -1100,
+	.max_voice_rx_vol[VOC_NB_INDEX] = 1000, //400,
+	.min_voice_rx_vol[VOC_NB_INDEX] = -500, //-1100,
+	.max_voice_rx_vol[VOC_WB_INDEX] = 1000, //400,
+	.min_voice_rx_vol[VOC_WB_INDEX] = -500, //-1100,
 };
 
 static struct snddev_ecodec_data snddev_bt_sco_mic_data = {
@@ -294,10 +294,10 @@ static struct snddev_icodec_data snddev_ihs_stereo_rx_data = {
 	.pamp_off = NULL,
 	.voltage_on = msm_snddev_hsed_voltage_on,
 	.voltage_off = msm_snddev_hsed_voltage_off,
-	.max_voice_rx_vol[VOC_NB_INDEX] = -700,
-	.min_voice_rx_vol[VOC_NB_INDEX] = -2200,
-	.max_voice_rx_vol[VOC_WB_INDEX] = -300,
-	.min_voice_rx_vol[VOC_WB_INDEX] = -1800,
+	.max_voice_rx_vol[VOC_NB_INDEX] = 600, //-700,
+	.min_voice_rx_vol[VOC_NB_INDEX] = -900, //-2200,
+	.max_voice_rx_vol[VOC_WB_INDEX] = 1000, //-300,
+	.min_voice_rx_vol[VOC_WB_INDEX] = -500, //-1800,
 };
 
 static struct platform_device msm_headset_stereo_device = {
@@ -694,10 +694,10 @@ static struct snddev_icodec_data snddev_ihs_ffa_mono_rx_data = {
 	.default_sample_rate = 48000,
 	.pamp_on = msm_snddev_hsed_voltage_on,
 	.pamp_off = msm_snddev_hsed_voltage_off,
-	.max_voice_rx_vol[VOC_NB_INDEX] = -700,
-	.min_voice_rx_vol[VOC_NB_INDEX] = -2200,
-	.max_voice_rx_vol[VOC_WB_INDEX] = -300,
-	.min_voice_rx_vol[VOC_WB_INDEX] = -1800,
+	.max_voice_rx_vol[VOC_NB_INDEX] = 600, //-700,
+	.min_voice_rx_vol[VOC_NB_INDEX] = -900, //-2200,
+	.max_voice_rx_vol[VOC_WB_INDEX] = 1000, //-300,
+	.min_voice_rx_vol[VOC_WB_INDEX] = -500, //-1800,
 };
 
 static struct platform_device msm_ihs_ffa_mono_rx_device = {
@@ -740,10 +740,10 @@ static struct snddev_icodec_data snddev_ihs_stereo_speaker_stereo_rx_data = {
 	.pamp_off = msm_snddev_poweramp_off,
 	.voltage_on = msm_snddev_hsed_voltage_on,
 	.voltage_off = msm_snddev_hsed_voltage_off,
-	.max_voice_rx_vol[VOC_NB_INDEX] = -500,
-	.min_voice_rx_vol[VOC_NB_INDEX] = -2000,
-	.max_voice_rx_vol[VOC_WB_INDEX] = -300,
-	.min_voice_rx_vol[VOC_WB_INDEX] = -1800,
+	.max_voice_rx_vol[VOC_NB_INDEX] = 800, //-500,
+	.min_voice_rx_vol[VOC_NB_INDEX] = -700, //-2000,
+	.max_voice_rx_vol[VOC_WB_INDEX] = 1000, //-300,
+	.min_voice_rx_vol[VOC_WB_INDEX] = -500, //-1800,
 };
 
 static struct platform_device msm_ihs_stereo_speaker_stereo_rx_device = {
