diff --git a/arch/arm/mach-msm/board-semc_mogami.c b/arch/arm/mach-msm/board-semc_mogami.c
index 3db2c9b..ede098d 100644
--- a/arch/arm/mach-msm/board-semc_mogami.c
+++ b/arch/arm/mach-msm/board-semc_mogami.c
@@ -191,7 +191,7 @@
 #define PMEM_KERNEL_EBI1_SIZE   0x600000
 
 #define PMIC_GPIO_INT		27
-#define PMIC_VREG_WLAN_LEVEL	2900
+#define PMIC_VREG_WLAN_LEVEL	2700
 #define PMIC_GPIO_SD_DET	22
 #ifdef CONFIG_PMIC_GPIO_25
 #define PMIC_GPIO_SD_POWER	25
@@ -2023,7 +2023,7 @@ static struct sii9024_platform_data sii9024_platform_data = {
 static void semc_mogami_lcd_regulators_on(void)
 {
 	vreg_helper_on("gp7",1800);  /* L8 */
-	vreg_helper_on("gp6",2850);  /* L15 */
+	vreg_helper_on("gp6",2300);  /* L15 */
 }
 
 /* Generic Power On function for SEMC mogami displays */
@@ -2805,7 +2805,7 @@ static struct bma250_platform_data bma250_platform_data = {
 #ifdef CONFIG_INPUT_APDS9702
 
 #define APDS9702_DOUT_GPIO   88
-#define APDS9702_VDD_VOLTAGE 2900
+#define APDS9702_VDD_VOLTAGE 2400
 #define APDS9702_WAIT_TIME   5000
 
 static int apds9702_gpio_setup(int request)
@@ -4250,7 +4250,7 @@ static uint32_t wifi_setup_power(struct device *dv, unsigned int vdd)
 }
 
 static struct mmc_platform_data msm7x30_sdc3_data = {
-	.ocr_mask = MMC_VDD_27_28 | MMC_VDD_28_29,
+	.ocr_mask = MMC_VDD_20_21 | MMC_VDD_21_22,
 	.translate_vdd = wifi_setup_power,
 	.mmc_bus_width = MMC_CAP_4_BIT_DATA | MMC_CAP_POWER_OFF_CARD,
 	.sdiowakeup_irq = MSM_GPIO_TO_INT(118),
@@ -4367,9 +4367,9 @@ static void __init mogami_temp_fixups(void)
 
 static void __init shared_vreg_on(void)
 {
-	vreg_helper_on(VREG_L20, 3050);
+	vreg_helper_on(VREG_L20, 2800);
 	vreg_helper_on(VREG_L10, 2600);
-	vreg_helper_on(VREG_L15, 2900);
+	vreg_helper_on(VREG_L15, 2300);
 	vreg_helper_on(VREG_L8, 1800);
 }
 
