diff --git a/arch/arm/mach-msm/leds-iyokan.c b/arch/arm/mach-msm/leds-iyokan.c
index 259101a..02bdd19 100644
--- a/arch/arm/mach-msm/leds-iyokan.c
+++ b/arch/arm/mach-msm/leds-iyokan.c
@@ -17,8 +17,8 @@ static struct as3676_platform_led as3676_leds_mapping[] = {
 		.sinks = BIT(AS3676_SINK_01),
 		.flags = AS3676_FLAG_PWM_CTRL | AS3676_FLAG_PWM_INIT
 			| AS3676_FLAG_WAIT_RESUME,
-		.max_current = 25000,
-		.default_brightness = LED_FULL,
+		.max_current = 30000,
+		.default_brightness = LED_FULL / 2,
 	},
 	{
 		.name = "button-backlight",
diff --git a/arch/arm/mach-msm/leds-mango.c b/arch/arm/mach-msm/leds-mango.c
index db623c3..70f8f61 100644
--- a/arch/arm/mach-msm/leds-mango.c
+++ b/arch/arm/mach-msm/leds-mango.c
@@ -17,8 +17,8 @@ static struct as3676_platform_led as3676_leds_mapping[] = {
 		.sinks = BIT(AS3676_SINK_01),
 		.flags = AS3676_FLAG_PWM_CTRL | AS3676_FLAG_PWM_INIT
 			| AS3676_FLAG_WAIT_RESUME,
-		.max_current = 25000,
-		.default_brightness = LED_FULL,
+		.max_current = 30000,
+		.default_brightness = LED_FULL / 2,
 	},
 	{
 		.name = "button-backlight",
