diff --git a/arch/arm/configs/nAa_anzu_defconfig b/arch/arm/configs/nAa_anzu_defconfig
index 0236fb5..de75e1a 100644
--- a/arch/arm/configs/nAa_anzu_defconfig
+++ b/arch/arm/configs/nAa_anzu_defconfig
@@ -423,6 +423,10 @@ CONFIG_MSM_SPM_V1=y
 # CONFIG_MSM_SPM_V2 is not set
 # CONFIG_MSM_RTB is not set
 # CONFIG_MSM_CACHE_DUMP is not set
+# CONFIG_MSM_UNDERVOLT_LCD is not set
+CONFIG_MSM_UNDERVOLT_PROXIMITY=y
+CONFIG_MSM_UNDERVOLT_WIFI=y
+CONFIG_MSM_UNDERVOLT_TOUCH=y
 
 #
 # System MMU
diff --git a/arch/arm/configs/nAa_coconut_defconfig b/arch/arm/configs/nAa_coconut_defconfig
index 3effa73..662be8e 100644
--- a/arch/arm/configs/nAa_coconut_defconfig
+++ b/arch/arm/configs/nAa_coconut_defconfig
@@ -423,6 +423,10 @@ CONFIG_MSM_SPM_V1=y
 # CONFIG_MSM_SPM_V2 is not set
 # CONFIG_MSM_RTB is not set
 # CONFIG_MSM_CACHE_DUMP is not set
+CONFIG_MSM_UNDERVOLT_LCD=y
+CONFIG_MSM_UNDERVOLT_PROXIMITY=y
+CONFIG_MSM_UNDERVOLT_WIFI=y
+CONFIG_MSM_UNDERVOLT_TOUCH=y
 
 #
 # System MMU
diff --git a/arch/arm/configs/nAa_haida_defconfig b/arch/arm/configs/nAa_haida_defconfig
index 1f558b9..5f7a27d 100644
--- a/arch/arm/configs/nAa_haida_defconfig
+++ b/arch/arm/configs/nAa_haida_defconfig
@@ -423,6 +423,10 @@ CONFIG_MSM_SPM_V1=y
 # CONFIG_MSM_SPM_V2 is not set
 # CONFIG_MSM_RTB is not set
 # CONFIG_MSM_CACHE_DUMP is not set
+CONFIG_MSM_UNDERVOLT_LCD=y
+CONFIG_MSM_UNDERVOLT_PROXIMITY=y
+CONFIG_MSM_UNDERVOLT_WIFI=y
+CONFIG_MSM_UNDERVOLT_TOUCH=y
 
 #
 # System MMU
diff --git a/arch/arm/configs/nAa_hallon_defconfig b/arch/arm/configs/nAa_hallon_defconfig
index 9423ddd..cef8a68 100644
--- a/arch/arm/configs/nAa_hallon_defconfig
+++ b/arch/arm/configs/nAa_hallon_defconfig
@@ -423,6 +423,10 @@ CONFIG_MSM_SPM_V1=y
 # CONFIG_MSM_SPM_V2 is not set
 # CONFIG_MSM_RTB is not set
 # CONFIG_MSM_CACHE_DUMP is not set
+CONFIG_MSM_UNDERVOLT_LCD=y
+CONFIG_MSM_UNDERVOLT_PROXIMITY=y
+CONFIG_MSM_UNDERVOLT_WIFI=y
+CONFIG_MSM_UNDERVOLT_TOUCH=y
 
 #
 # System MMU
diff --git a/arch/arm/configs/nAa_iyokan_defconfig b/arch/arm/configs/nAa_iyokan_defconfig
index e32613f..35518cd 100644
--- a/arch/arm/configs/nAa_iyokan_defconfig
+++ b/arch/arm/configs/nAa_iyokan_defconfig
@@ -423,6 +423,10 @@ CONFIG_MSM_SPM_V1=y
 # CONFIG_MSM_SPM_V2 is not set
 # CONFIG_MSM_RTB is not set
 # CONFIG_MSM_CACHE_DUMP is not set
+CONFIG_MSM_UNDERVOLT_LCD=y
+CONFIG_MSM_UNDERVOLT_PROXIMITY=y
+CONFIG_MSM_UNDERVOLT_WIFI=y
+CONFIG_MSM_UNDERVOLT_TOUCH=y
 
 #
 # System MMU
diff --git a/arch/arm/configs/nAa_mango_defconfig b/arch/arm/configs/nAa_mango_defconfig
index cb17e90..4ece804 100644
--- a/arch/arm/configs/nAa_mango_defconfig
+++ b/arch/arm/configs/nAa_mango_defconfig
@@ -423,6 +423,10 @@ CONFIG_MSM_SPM_V1=y
 # CONFIG_MSM_SPM_V2 is not set
 # CONFIG_MSM_RTB is not set
 # CONFIG_MSM_CACHE_DUMP is not set
+CONFIG_MSM_UNDERVOLT_LCD=y
+CONFIG_MSM_UNDERVOLT_PROXIMITY=y
+CONFIG_MSM_UNDERVOLT_WIFI=y
+CONFIG_MSM_UNDERVOLT_TOUCH=y
 
 #
 # System MMU
diff --git a/arch/arm/configs/nAa_satsuma_defconfig b/arch/arm/configs/nAa_satsuma_defconfig
index 7cf325c..aa7d615 100644
--- a/arch/arm/configs/nAa_satsuma_defconfig
+++ b/arch/arm/configs/nAa_satsuma_defconfig
@@ -423,6 +423,10 @@ CONFIG_MSM_SPM_V1=y
 # CONFIG_MSM_SPM_V2 is not set
 # CONFIG_MSM_RTB is not set
 # CONFIG_MSM_CACHE_DUMP is not set
+CONFIG_MSM_UNDERVOLT_LCD=y
+CONFIG_MSM_UNDERVOLT_PROXIMITY=y
+CONFIG_MSM_UNDERVOLT_WIFI=y
+CONFIG_MSM_UNDERVOLT_TOUCH=y
 
 #
 # System MMU
diff --git a/arch/arm/configs/nAa_smultron_defconfig b/arch/arm/configs/nAa_smultron_defconfig
index dc1cfa3..7243c12 100644
--- a/arch/arm/configs/nAa_smultron_defconfig
+++ b/arch/arm/configs/nAa_smultron_defconfig
@@ -423,6 +423,10 @@ CONFIG_MSM_SPM_V1=y
 # CONFIG_MSM_SPM_V2 is not set
 # CONFIG_MSM_RTB is not set
 # CONFIG_MSM_CACHE_DUMP is not set
+CONFIG_MSM_UNDERVOLT_LCD=y
+CONFIG_MSM_UNDERVOLT_PROXIMITY=y
+CONFIG_MSM_UNDERVOLT_WIFI=y
+CONFIG_MSM_UNDERVOLT_TOUCH=y
 
 #
 # System MMU
diff --git a/arch/arm/configs/nAa_urushi_defconfig b/arch/arm/configs/nAa_urushi_defconfig
index e7a5d93..3a950f3 100644
--- a/arch/arm/configs/nAa_urushi_defconfig
+++ b/arch/arm/configs/nAa_urushi_defconfig
@@ -423,6 +423,10 @@ CONFIG_MSM_SPM_V1=y
 # CONFIG_MSM_SPM_V2 is not set
 # CONFIG_MSM_RTB is not set
 # CONFIG_MSM_CACHE_DUMP is not set
+CONFIG_MSM_UNDERVOLT_LCD=y
+CONFIG_MSM_UNDERVOLT_PROXIMITY=y
+CONFIG_MSM_UNDERVOLT_WIFI=y
+CONFIG_MSM_UNDERVOLT_TOUCH=y
 
 #
 # System MMU
diff --git a/arch/arm/mach-msm/Kconfig b/arch/arm/mach-msm/Kconfig
index 042098b..7c23ce9 100644
--- a/arch/arm/mach-msm/Kconfig
+++ b/arch/arm/mach-msm/Kconfig
@@ -477,6 +477,10 @@ config MACH_SEMC_MOGAMI
 config MACH_SEMC_SMULTRON
 	depends on ARCH_MSM7X30
 	depends on !MSM_STACKED_MEMORY
+	select MSM_UNDERVOLT_LCD
+	select MSM_UNDERVOLT_PROXIMITY
+	select MSM_UNDERVOLT_WIFI
+	select MSM_UNDERVOLT_TOUCH
 	select SUPPORT_ALIEN_USB_CHARGER
 	default y
 	bool "SonyEricsson Smultron"
@@ -486,6 +490,10 @@ config MACH_SEMC_SMULTRON
 config MACH_SEMC_MANGO
 	depends on ARCH_MSM7X30
 	depends on !MSM_STACKED_MEMORY
+	select MSM_UNDERVOLT_LCD
+	select MSM_UNDERVOLT_PROXIMITY
+	select MSM_UNDERVOLT_WIFI
+	select MSM_UNDERVOLT_TOUCH
 	select SUPPORT_ALIEN_USB_CHARGER
 	default n
 	bool "SonyEricsson Mango"
@@ -495,6 +503,10 @@ config MACH_SEMC_MANGO
 config MACH_SEMC_COCONUT
 	depends on ARCH_MSM7X30
 	depends on !MSM_STACKED_MEMORY
+	select MSM_UNDERVOLT_LCD
+	select MSM_UNDERVOLT_PROXIMITY
+	select MSM_UNDERVOLT_WIFI
+	select MSM_UNDERVOLT_TOUCH
 	select SUPPORT_ALIEN_USB_CHARGER
 	default n
 	bool "SonyEricsson Coconut"
@@ -505,6 +517,10 @@ config MACH_SEMC_HAIDA
 	depends on ARCH_MSM7X30
 	depends on !MSM_STACKED_MEMORY
 	select FB_MSM_HDPI
+	select MSM_UNDERVOLT_LCD
+	select MSM_UNDERVOLT_PROXIMITY
+	select MSM_UNDERVOLT_WIFI
+	select MSM_UNDERVOLT_TOUCH
 	select SUPPORT_ALIEN_USB_CHARGER
 	default y
 	bool "SonyEricsson Haida"
@@ -515,6 +531,10 @@ config MACH_SEMC_URUSHI
 	depends on ARCH_MSM7X30
 	depends on !MSM_STACKED_MEMORY
 	select FB_MSM_HDPI
+	select MSM_UNDERVOLT_LCD
+	select MSM_UNDERVOLT_PROXIMITY
+	select MSM_UNDERVOLT_WIFI
+	select MSM_UNDERVOLT_TOUCH
 	select SUPPORT_ALIEN_USB_CHARGER
 	default y
 	bool "SonyEricsson Urushi"
@@ -525,6 +545,9 @@ config MACH_SEMC_ANZU
 	depends on ARCH_MSM7X30
 	depends on !MSM_STACKED_MEMORY
 	select FB_MSM_HDPI
+	select MSM_UNDERVOLT_PROXIMITY
+	select MSM_UNDERVOLT_WIFI
+	select MSM_UNDERVOLT_TOUCH
 	select SUPPORT_ALIEN_USB_CHARGER
 	default y
 	bool "SonyEricsson Anzu"
@@ -534,6 +557,10 @@ config MACH_SEMC_ANZU
 config MACH_SEMC_IYOKAN
 	depends on ARCH_MSM7X30
 	select FB_MSM_HDPI
+	select MSM_UNDERVOLT_LCD
+	select MSM_UNDERVOLT_PROXIMITY
+	select MSM_UNDERVOLT_WIFI
+	select MSM_UNDERVOLT_TOUCH
 	depends on !MSM_STACKED_MEMORY
 	select SUPPORT_ALIEN_USB_CHARGER
 	default y
@@ -545,6 +572,10 @@ config MACH_SEMC_HALLON
 	depends on ARCH_MSM7X30
 	depends on !MSM_STACKED_MEMORY
 	select FB_MSM_HDPI
+	select MSM_UNDERVOLT_LCD
+	select MSM_UNDERVOLT_PROXIMITY
+	select MSM_UNDERVOLT_WIFI
+	select MSM_UNDERVOLT_TOUCH
 	select SUPPORT_ALIEN_USB_CHARGER
 	default y
 	bool "SonyEricsson Hallon"
@@ -554,6 +585,10 @@ config MACH_SEMC_HALLON
 config MACH_SEMC_SATSUMA
 	depends on ARCH_MSM7X30
 	depends on !MSM_STACKED_MEMORY
+	select MSM_UNDERVOLT_LCD
+	select MSM_UNDERVOLT_PROXIMITY
+	select MSM_UNDERVOLT_WIFI
+	select MSM_UNDERVOLT_TOUCH
 	select SUPPORT_ALIEN_USB_CHARGER
 	default y
 	bool "SonyEricsson Satsuma"
@@ -2262,6 +2297,26 @@ config MSM_CACHE_DUMP
 	  This allows for analysis of the caches in case cache corruption is
 	  suspected.
 
+config MSM_UNDERVOLT_LCD
+	bool "Undervolt LCD"
+	help
+	  Undervolt LCD from 2850 to 2300 mV to reduce energy consumption
+
+config MSM_UNDERVOLT_PROXIMITY
+	bool "Undervolt proximity sensor"
+	help
+	  Undervolt proximity sensor from 2900 to 2400 mV to reduce energy consumption
+
+config MSM_UNDERVOLT_WIFI
+	bool "Undervolt Wi-Fi"
+	help
+	  Undervolt Wi-Fi from 2750/2850 to 2050/2150 mV to reduce energy consumption
+
+config MSM_UNDERVOLT_TOUCH
+	bool "Undervolt touch panel"
+	help
+	  Undervolt touch panel from 3050 to 2800 mV to reduce energy consumption
+
 config MSM_DCVS
 	bool "Use MSM DCVS for CPU/GPU Frequency control"
 	depends on MSM_SCM
diff --git a/arch/arm/mach-msm/board-semc_mogami.c b/arch/arm/mach-msm/board-semc_mogami.c
index f3a631f..fe5cb93 100644
--- a/arch/arm/mach-msm/board-semc_mogami.c
+++ b/arch/arm/mach-msm/board-semc_mogami.c
@@ -3227,10 +3227,10 @@ static struct sii9024_platform_data sii9024_platform_data = {
 static void semc_mogami_lcd_regulators_on(void)
 {
 	vreg_helper_on("gp7",1800);  /* L8 */
-#ifdef CONFIG_MACH_SEMC_ANZU
-	vreg_helper_on("gp6",2850);  /* L15 */
-#else
+#ifdef CONFIG_MSM_UNDERVOLT_LCD
 	vreg_helper_on("gp6",2300);  /* L15 */
+#else
+	vreg_helper_on("gp6",2850);  /* L15 */
 #endif
 }
 
@@ -4132,7 +4132,11 @@ static struct bma250_platform_data bma250_platform_data = {
 
 #ifdef CONFIG_INPUT_APDS9702
 #define APDS9702_DOUT_GPIO   88
+#ifdef CONFIG_MSM_UNDERVOLT_PROXIMITY_SENSOR
 #define APDS9702_VDD_VOLTAGE 2400
+#else
+#define APDS9702_VDD_VOLTAGE 2900
+#endif
 #define APDS9702_WAIT_TIME   5000
 
 static int apds9702_gpio_setup(int request)
@@ -6601,7 +6605,11 @@ static uint32_t wifi_setup_power(struct device *dv, unsigned int vdd)
 }
 
 static struct mmc_platform_data msm7x30_sdc3_data = {
+#ifdef CONFIG_MSM_UNDERVOLT_WIFI
 	.ocr_mask	= MMC_VDD_20_21 | MMC_VDD_21_22,
+#else
+	.ocr_mask	= MMC_VDD_27_28 | MMC_VDD_28_29,
+#endif
 	.translate_vdd	= wifi_setup_power,
 	.mmc_bus_width  = MMC_CAP_4_BIT_DATA | MMC_CAP_POWER_OFF_CARD,
 	.sdiowakeup_irq = MSM_GPIO_TO_INT(118),
@@ -6785,12 +6793,16 @@ static void __init mogami_temp_fixups(void)
 
 static void __init shared_vreg_on(void)
 {
+#ifdef CONFIG_MSM_UNDERVOLT_TOUCH
 	vreg_helper_on(VREG_L20, 2800);
-	vreg_helper_on(VREG_L10, 2600);
-#ifdef CONFIG_MACH_SEMC_ANZU
-	vreg_helper_on(VREG_L15, 2900);
 #else
+	vreg_helper_on(VREG_L20, 3050);
+#endif
+	vreg_helper_on(VREG_L10, 2600);
+#ifdef CONFIG_MSM_UNDERVOLT_LCD
 	vreg_helper_on(VREG_L15, 2300);
+#else
+	vreg_helper_on(VREG_L15, 2900);
 #endif
 	vreg_helper_on(VREG_L8, 1800);
 }
