diff --git a/arch/arm/mach-msm/qdsp5v2/snddev_ecodec.c b/arch/arm/mach-msm/qdsp5v2/snddev_ecodec.c
index 1bab06c..c122702 100644
--- a/arch/arm/mach-msm/qdsp5v2/snddev_ecodec.c
+++ b/arch/arm/mach-msm/qdsp5v2/snddev_ecodec.c
@@ -27,6 +27,8 @@
 #include <mach/qdsp5v2/afe.h>
 #include <mach/debug_mm.h>
 
+static int marimba_eoff = 750;
+
 /* Context for each external codec device */
 struct snddev_ecodec_state {
 	struct snddev_ecodec_data *data;
@@ -349,10 +351,19 @@ static int snddev_ecodec_probe(struct platform_device *pdev)
 	ecodec->sample_rate = 8000; /* Default to 8KHz */
 	 if (pdata->capability & SNDDEV_CAP_RX) {
 		for (i = 0; i < VOC_RX_VOL_ARRAY_NUM; i++) {
+			printk(KERN_WARNING "Marimba vol %s/%d %d..%d %d\n",
+				pdata->name, i,
+				pdata->min_voice_rx_vol[i],
+				pdata->max_voice_rx_vol[i],
+				marimba_eoff);
 			dev_info->max_voc_rx_vol[i] =
 				pdata->max_voice_rx_vol[i];
 			dev_info->min_voc_rx_vol[i] =
 				pdata->min_voice_rx_vol[i];
+			if (dev_info->max_voc_rx_vol[i] != dev_info->min_voc_rx_vol[i]) {
+				dev_info->max_voc_rx_vol[i] += marimba_eoff;
+				dev_info->min_voc_rx_vol[i] += marimba_eoff;
+			}
 		}
 	}
 error:
diff --git a/arch/arm/mach-msm/qdsp5v2/snddev_icodec.c b/arch/arm/mach-msm/qdsp5v2/snddev_icodec.c
index 5c23504..ed25fcb 100644
--- a/arch/arm/mach-msm/qdsp5v2/snddev_icodec.c
+++ b/arch/arm/mach-msm/qdsp5v2/snddev_icodec.c
@@ -36,6 +36,8 @@
 #include <mach/rpc_pmapp.h>
 #include <mach/qdsp5v2/audio_acdb_def.h>
 
+static int marimba_ioff = 750;
+
 #define SMPS_AUDIO_PLAYBACK_ID	"AUPB"
 #define SMPS_AUDIO_RECORD_ID	"AURC"
 
@@ -772,10 +774,19 @@ static int snddev_icodec_probe(struct platform_device *pdev)
 	dev_info->sample_rate = pdata->default_sample_rate;
 	if (pdata->capability & SNDDEV_CAP_RX) {
 		for (i = 0; i < VOC_RX_VOL_ARRAY_NUM; i++) {
+			printk(KERN_WARNING "Marimba vol %s/%d %d..%d %d\n",
+				pdata->name, i,
+				pdata->min_voice_rx_vol[i],
+				pdata->max_voice_rx_vol[i],
+				marimba_ioff);
 			dev_info->max_voc_rx_vol[i] =
 				pdata->max_voice_rx_vol[i];
 			dev_info->min_voc_rx_vol[i] =
 				pdata->min_voice_rx_vol[i];
+			if (dev_info->max_voc_rx_vol[i] != dev_info->min_voc_rx_vol[i]) {
+				dev_info->max_voc_rx_vol[i] += marimba_ioff;
+				dev_info->min_voc_rx_vol[i] += marimba_ioff;
+			}
 		}
 		dev_info->dev_ops.enable_sidetone =
 		snddev_icodec_enable_sidetone;
