diff --git a/src/com/android/mms/data/WorkingMessage.java b/src/com/android/mms/data/WorkingMessage.java
index 14dd7b4..7510b00 100755
--- a/src/com/android/mms/data/WorkingMessage.java
+++ b/src/com/android/mms/data/WorkingMessage.java
@@ -1476,7 +1476,7 @@ public class WorkingMessage {
 
         Uri uri;
         try {
-            if (cursor.moveToFirst()) {
+            if (cursor != null && cursor.moveToFirst()) {
                 uri = ContentUris.withAppendedId(Mms.Draft.CONTENT_URI,
                         cursor.getLong(MMS_ID_INDEX));
                 String subject = MessageUtils.extractEncStrFromCursor(cursor, MMS_SUBJECT_INDEX,
@@ -1490,7 +1490,8 @@ public class WorkingMessage {
                 return uri;
             }
         } finally {
-            cursor.close();
+            if (cursor != null)
+                cursor.close();
         }
 
         return null;
