diff --git a/prebuilt/init.semc.rc b/prebuilt/init.semc.rc
index d0d1e98..11eb089 100644
--- a/prebuilt/init.semc.rc
+++ b/prebuilt/init.semc.rc
@@ -117,6 +117,24 @@ on boot
     write /sys/devices/system/cpu/cpufreq/ondemand/sampling_down_factor 4
     write /sys/devices/system/cpu/cpufreq/ondemand/down_differential 10
 
+    # Xtended
+    write /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor smartassV2
+    write /sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq 1017600
+    write /sys/devices/system/cpu/cpu0/cpufreq/scaling_min_freq 249600
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/awake_ideal_freq
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/sleep_ideal_freq
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/sleep_wakeup_freq
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/max_cpu_load
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/min_cpu_load
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/ramp_up_step
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/ramp_down_step
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/up_rate_us
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/down_rate_us
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/boost_enabled
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/boost_pulse
+    chown root system /sys/devices/system/cpu/cpu0/cpufreq/vdd_levels
+    chown root system /sys/module/lowmemorykiller/parameters/minfree
+
     mount debugfs /sys/kernel/debug /sys/kernel/debug
 
 on property:init.svc.bluetoothd=running
