diff --git a/prebuilt/init.semc.rc b/prebuilt/init.semc.rc
index d0d1e98..b7a1fe5 100644
--- a/prebuilt/init.semc.rc
+++ b/prebuilt/init.semc.rc
@@ -117,6 +117,58 @@ on boot
     write /sys/devices/system/cpu/cpufreq/ondemand/sampling_down_factor 4
     write /sys/devices/system/cpu/cpufreq/ondemand/down_differential 10
 
+    # Xtended
+    write /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor smartassV2
+
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/awake_ideal_freq
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/sleep_ideal_freq
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/sleep_wakeup_freq
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/max_cpu_load
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/min_cpu_load
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/ramp_up_step
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/ramp_down_step
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/up_rate_us
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/down_rate_us
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/boost_enabled
+    chown root system /sys/devices/system/cpu/cpufreq/smartass/boost_pulse
+
+    chmod 664 /sys/devices/system/cpu/cpufreq/smartass/awake_ideal_freq
+    chmod 664 /sys/devices/system/cpu/cpufreq/smartass/sleep_ideal_freq
+    chmod 664 /sys/devices/system/cpu/cpufreq/smartass/sleep_wakeup_freq
+    chmod 664 /sys/devices/system/cpu/cpufreq/smartass/max_cpu_load
+    chmod 664 /sys/devices/system/cpu/cpufreq/smartass/min_cpu_load
+    chmod 664 /sys/devices/system/cpu/cpufreq/smartass/ramp_up_step
+    chmod 664 /sys/devices/system/cpu/cpufreq/smartass/ramp_down_step
+    chmod 664 /sys/devices/system/cpu/cpufreq/smartass/up_rate_us
+    chmod 664 /sys/devices/system/cpu/cpufreq/smartass/down_rate_us
+    chmod 664 /sys/devices/system/cpu/cpufreq/smartass/boost_enabled
+    chmod 664 /sys/devices/system/cpu/cpufreq/smartass/boost_pulse
+
+    write /sys/devices/system/cpu/cpufreq/smartass/awake_ideal_freq 768000
+    write /sys/devices/system/cpu/cpufreq/smartass/sleep_ideal_freq 134400
+    write /sys/devices/system/cpu/cpufreq/smartass/sleep_wakeup_freq 249600
+    write /sys/devices/system/cpu/cpufreq/smartass/max_cpu_load 50
+    write /sys/devices/system/cpu/cpufreq/smartass/min_cpu_load 25
+    write /sys/devices/system/cpu/cpufreq/smartass/ramp_up_step 128000
+    write /sys/devices/system/cpu/cpufreq/smartass/ramp_down_step 128000
+    write /sys/devices/system/cpu/cpufreq/smartass/up_rate_us 48000
+    write /sys/devices/system/cpu/cpufreq/smartass/down_rate_us 48000
+    write /sys/devices/system/cpu/cpufreq/smartass/boost_enabled 0
+
+    chown root system /sys/devices/system/cpu/cpu0/cpufreq/vdd_levels
+    chmod 664 /sys/devices/system/cpu/cpu0/cpufreq/vdd_levels
+
+    chown root system /sys/module/lowmemorykiller/parameters/minfree
+    chmod 664 /sys/module/lowmemorykiller/parameters/minfree
+
+    chown root system /sys/class/leds/lcd-backlight/als/enable
+    chmod 664 /sys/class/leds/lcd-backlight/als/enable
+
+    #/proc/sys/kernel/sched_autogroup_enabled
+
+    chown root system /sys/module/sync/parameters/fsync_enabled
+    chmod 664 /sys/module/sync/parameters/fsync_enabled
+
     mount debugfs /sys/kernel/debug /sys/kernel/debug
 
 on property:init.svc.bluetoothd=running
