--- frameworks/base/media/java/android/media/MediaRecorder.java	2012-12-08 13:10:55.000000000 +0100
+++ frameworks/base/media/java/android/media/MediaRecorder.java	2012-12-08 13:11:19.000000000 +0100
@@ -1128,6 +1128,7 @@ public class MediaRecorder
 				pSetMan.notification(packageName[0], 0, PrivacySettings.REAL, PrivacySettings.DATA_CAMERA, null, null);
 			else if(packageName != null && packageName.length > 0)
 				pSetMan.notification(packageName[0], 0, PrivacySettings.REAL, PrivacySettings.DATA_RECORD_AUDIO, null, null);
+			deletedFile = true;
 		}
 
         if (mPath != null) {
