diff --git a/combinedroot/init.sh b/combinedroot/init.sh
index e4dda96..bfd8de3 100644
--- a/combinedroot/init.sh
+++ b/combinedroot/init.sh
@@ -53,7 +53,7 @@ busybox sleep 3
 load_image=/sbin/ramdisk.cpio
 
 # boot decision
-if [ -s /dev/keycheck -o -e /cache/recovery/boot ]
+if [ -s /dev/keycheck -o -e /cache/recovery/boot -o ! -e /cache/recovery/pinok ]
 then
 	busybox echo 'RECOVERY BOOT' >>boot.txt
 	busybox rm -fr /cache/recovery/boot
@@ -69,6 +69,10 @@ then
 	load_image=/sbin/ramdisk-recovery.cpio
 else
 	busybox echo 'ANDROID BOOT' >>boot.txt
+	if [ -e /cache/recovery/pincode ]
+	then
+		busybox rm -fr /cache/recovery/pinok
+	fi
 	# poweroff LED & button-backlight
 	busybox echo 0 > ${BOOTREC_LED_RED}/brightness
 	busybox echo 0 > ${BOOTREC_LED_GREEN}/brightness
