diff --git a/init/init.c b/init/init.c
index fe7dbba..550fd58 100755
--- a/init/init.c
+++ b/init/init.c
@@ -673,6 +673,8 @@ static void import_kernel_nv(char *name, int for_emulator)
         cnt = snprintf(prop, sizeof(prop), "ro.boot.%s", boot_prop_name);
         if (cnt < PROP_NAME_MAX)
             property_set(prop, value);
+    } else if (!strcmp(name,"serialno")) {
+        property_set("ro.boot.serialno", value);
     }
 }
 
