diff --git a/partitionmanager.cpp b/partitionmanager.cpp
index 2614383..44067e4 100644
--- a/partitionmanager.cpp
+++ b/partitionmanager.cpp
@@ -1506,6 +1506,8 @@ void TWPartitionManager::Update_System_Details(void) {
 #endif
 		}
 	}
+	DataManager::SetValue(TW_HAS_ANDROID_SECURE, 1);
+	DataManager::SetValue(TW_BACKUP_BOOT_VAR, 0);
 	DataManager::SetValue(TW_BACKUP_DATA_SIZE, data_size);
 	string current_storage_path = DataManager::GetCurrentStoragePath();
 	TWPartition* FreeStorage = Find_Partition_By_Path(current_storage_path);
diff --git a/prebuilt/Android.mk b/prebuilt/Android.mk
index e260598..902d9ea 100644
--- a/prebuilt/Android.mk
+++ b/prebuilt/Android.mk
@@ -85,6 +85,15 @@ LOCAL_MODULE_PATH := $(TARGET_RECOVERY_ROOT_OUT)/sbin
 LOCAL_SRC_FILES := $(LOCAL_MODULE)
 include $(BUILD_PREBUILT)
 
+#rebootrecovery
+include $(CLEAR_VARS)
+LOCAL_MODULE := rebootrecovery.sh
+LOCAL_MODULE_TAGS := eng
+LOCAL_MODULE_CLASS := RECOVERY_EXECUTABLES
+LOCAL_MODULE_PATH := $(TARGET_RECOVERY_ROOT_OUT)/sbin
+LOCAL_SRC_FILES := $(LOCAL_MODULE)
+include $(BUILD_PREBUILT)
+
 #mke2fs.conf
 include $(CLEAR_VARS)
 LOCAL_MODULE := mke2fs.conf
diff --git a/prebuilt/rebootrecovery.sh b/prebuilt/rebootrecovery.sh
new file mode 100644
index 0000000..11810aa
--- /dev/null
+++ b/prebuilt/rebootrecovery.sh
@@ -0,0 +1,2 @@
+#/sbin/sh
+touch /cache/recovery/boot
