diff --git a/twrp-functions.cpp b/twrp-functions.cpp
index a90be5a..d657e29 100644
--- a/twrp-functions.cpp
+++ b/twrp-functions.cpp
@@ -395,11 +395,13 @@ void TWFunc::check_and_run_script(const char* script_file, const char* display_n
 {
 	// Check for and run startup script if script exists
 	struct stat st;
-	string result;
 	if (stat(script_file, &st) == 0) {
 		ui_print("Running %s script...\n", display_name);
-		chmod(script_file, S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH);
-		TWFunc::Exec_Cmd(script_file, result);
+		char command[255];
+		strcpy(command, "chmod 755 ");
+		strcat(command, script_file);
+		__system(command);
+		__system(script_file);
 		ui_print("\nFinished running %s script.\n", display_name);
 	}
 }
diff --git a/prebuilt/Android.mk b/prebuilt/Android.mk
index dec000f..76ca1c9 100644
--- a/prebuilt/Android.mk
+++ b/prebuilt/Android.mk
@@ -96,6 +96,15 @@ LOCAL_MODULE_PATH := $(TARGET_RECOVERY_ROOT_OUT)/sbin
 LOCAL_SRC_FILES := $(LOCAL_MODULE)
 include $(BUILD_PREBUILT)
 
+#rebootrecovery
+include $(CLEAR_VARS)
+LOCAL_MODULE := rebootrecovery.sh
+LOCAL_MODULE_TAGS := eng
+LOCAL_MODULE_CLASS := RECOVERY_EXECUTABLES
+LOCAL_MODULE_PATH := $(TARGET_RECOVERY_ROOT_OUT)/sbin
+LOCAL_SRC_FILES := $(LOCAL_MODULE)
+include $(BUILD_PREBUILT)
+
 #mke2fs.conf
 include $(CLEAR_VARS)
 LOCAL_MODULE := mke2fs.conf
diff --git a/prebuilt/rebootrecovery.sh b/prebuilt/rebootrecovery.sh
new file mode 100644
index 0000000..11810aa
--- /dev/null
+++ b/prebuilt/rebootrecovery.sh
@@ -0,0 +1,2 @@
+#/sbin/sh
+touch /cache/recovery/boot
